File: app/models/project.rb
class Project < ActiveRecord::Base
  has_many :todos, -> {order("todos.due IS NULL, todos.due ASC, todos.created_at ASC")}, dependent: :delete_all
  has_many :notes, -> {order "created_at DESC"}, dependent: :delete_all
  has_many :recurring_todos

  belongs_to :default_context, :class_name => "Context", :foreign_key => "default_context_id"
  belongs_to :user

  scope :active,      -> { where state: 'active' }
  scope :hidden,      -> { where state: 'hidden' }
  scope :completed,   -> { where state: 'completed' }
  scope :uncompleted, -> { where("NOT(state = ?)", 'completed') }

  scope :with_name_or_description, lambda { |body| where("name LIKE ? OR description LIKE ?", body, body) }
  scope :with_namepart, lambda { |body| where("name LIKE ?", body + '%') }

  before_create :set_last_reviewed_now

  validates_presence_of :name
  validates_length_of :name, :maximum => 255
  validates_uniqueness_of :name, :scope => "user_id"

  acts_as_list :scope => 'user_id = #{user_id} AND state = \'#{state}\'', :top_of_list => 0

  include AASM

  aasm :column => :state do

    state :active, :initial => true
    state :hidden, :enter => :hide_todos, :exit => :unhide_todos
    state :completed, :enter => :set_completed_at_date, :exit => :clear_completed_at_date

    event :activate do
      transitions :to => :active,   :from => [:active, :hidden, :completed]
    end

    event :hide do
      transitions :to => :hidden,   :from => [:active, :completed]
    end

    event :complete do
      transitions :to => :completed, :from => [:active, :hidden]
    end
  end

  attr_accessor :cached_note_count

  def self.null_object
    NullProject.new
  end

  def set_last_reviewed_now
    self.last_reviewed = Time.now
  end

  def hide_todos
    todos.each do |t|
      unless t.completed? || t.deferred?
        t.hide!
        t.save
      end
    end
  end

  def unhide_todos
    todos.each do |t|
      if t.project_hidden?
        t.unhide!
        t.save
      end
    end
  end

  def set_completed_at_date
    self.completed_at = Time.zone.now
  end

  def clear_completed_at_date
    self.completed_at = nil
  end

  def note_count
    # TODO: test this for eager and not eager loading!!!
    return 0 if notes.size == 0
    cached_note_count || notes.count
  end

  alias_method :original_default_context, :default_context

  def default_context
    original_default_context.nil? ? Context.null_object : original_default_context
  end

  # would prefer to call this method state=(), but that causes an endless loop
  # as a result of acts_as_state_machine calling state=() to update the attribute
  def transition_to(candidate_state)
    case candidate_state.to_sym
      when aasm.current_state
        return
      when :hidden
        hide!
      when :active
        activate!
      when :completed
        complete!
    end
  end

  def needs_review?(user)
    return active? && ( last_reviewed.nil? ||
                        (last_reviewed < Time.current - user.prefs.review_period.days))
  end

  def blocked?
    ## mutually exclusive for stalled and blocked
    # blocked is uncompleted project with deferred or pending todos, but no next actions
    return false if self.completed?
    return !self.todos.deferred_or_blocked.empty? && self.todos.active.empty?
  end

  def stalled?
    # Stalled projects are active projects with no active next actions
    return false if self.completed? || self.hidden?
    return self.todos.deferred_or_blocked.empty? && self.todos.active.empty?
  end

  def shortened_name(length=40)
    name.truncate(length, :omission => "...").html_safe
  end

  def name=(value)
    if value
      self[:name] = value.gsub(/\s{2,}/, " ").strip
    else
      self[:name] = nil
    end
  end

  def new_record_before_save?
    @new_record_before_save
  end

  def age_in_days
    @age_in_days ||= (Time.current.to_date - created_at.to_date).to_i + 1
  end

  def running_time
    if completed_at.nil?
      return age_in_days
    else
      return (completed_at.to_date - created_at.to_date).to_i + 1
    end
  end

  def self.import(filename, params, user)
    count = 0
    CSV.foreach(filename, headers: true) do |row|
      unless find_by_name_and_user_id row[params[:name].to_i], user.id
        project = new
        project.name = row[params[:name].to_i]
        project.user = user
        project.description = row[params[:description].to_i] if row[params[:description].to_i].present?
        project.state = 'active'
        project.save!
        count += 1
      end
    end
    count
  end

end

class NullProject

  def hidden?
    false
  end

  def nil?
    true
  end

  def id
    nil
  end

  def name
    ""
  end

  def persisted?
    false
  end

end

File: internal/models/project.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// ProjectState represents the state of a project
type ProjectState string

const (
	ProjectStateActive    ProjectState = "active"
	ProjectStateHidden    ProjectState = "hidden"
	ProjectStateCompleted ProjectState = "completed"
)

// Project represents a GTD project
type Project struct {
	ID               uint           `gorm:"primaryKey" json:"id"`
	UserID           uint           `gorm:"not null;index" json:"user_id"`
	Name             string         `gorm:"not null;size:255" json:"name"`
	Description      string         `gorm:"type:text" json:"description"`
	Position         int            `gorm:"default:1" json:"position"`
	State            ProjectState   `gorm:"type:varchar(20);default:'active'" json:"state"`
	DefaultContextID *uint          `json:"default_context_id"`
	DefaultTags      string         `gorm:"type:text" json:"default_tags"`
	CompletedAt      *time.Time     `json:"completed_at"`
	LastReviewedAt   *time.Time     `json:"last_reviewed_at"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`

	// Associations
	User           User            `gorm:"foreignKey:UserID" json:"-"`
	DefaultContext *Context        `gorm:"foreignKey:DefaultContextID" json:"default_context,omitempty"`
	Todos          []Todo          `gorm:"foreignKey:ProjectID" json:"todos,omitempty"`
	RecurringTodos []RecurringTodo `gorm:"foreignKey:ProjectID" json:"recurring_todos,omitempty"`
	Notes          []Note          `gorm:"foreignKey:ProjectID" json:"notes,omitempty"`
}

// BeforeCreate sets default values
func (p *Project) BeforeCreate(tx *gorm.DB) error {
	if p.State == "" {
		p.State = ProjectStateActive
	}
	if p.Position == 0 {
		p.Position = 1
	}
	return nil
}

// IsActive returns true if the project is active
func (p *Project) IsActive() bool {
	return p.State == ProjectStateActive
}

// IsHidden returns true if the project is hidden
func (p *Project) IsHidden() bool {
	return p.State == ProjectStateHidden
}

// IsCompleted returns true if the project is completed
func (p *Project) IsCompleted() bool {
	return p.State == ProjectStateCompleted
}

// Hide sets the project state to hidden
func (p *Project) Hide() {
	p.State = ProjectStateHidden
}

// Activate sets the project state to active
func (p *Project) Activate() {
	p.State = ProjectStateActive
}

// Complete sets the project state to completed
func (p *Project) Complete() {
	now := time.Now()
	p.State = ProjectStateCompleted
	p.CompletedAt = &now
}

// MarkReviewed updates the last_reviewed_at timestamp
func (p *Project) MarkReviewed() {
	now := time.Now()
	p.LastReviewedAt = &now
}
